### DISCUSSION TITLE
Does LXQt DE have a bug that convolutes `ktimer`? Or the other way round?

### DISCUSSION THREAD

**Discussion by rajibando** [lxqt/lxqt#2665] [Q&A]
May please the discussion, [Post 9: How to make ktimer play wav files?](https://discuss.kde.org/t/how-to-make-ktimer-play-wav-files/27405/9), be perused.
It appears that the LXQt has a bug that convolutes the file browser interface of `ktimer`. Or there isn't, but is the other way round? Beyond my cognitive horizon.

 I have mentioned in Post No.6 therein that:
> The version of ktimer in my system is 4:22.12.3-1.

I already have a workaround to the current problem by having the `Command line` open a terminal and then execute the necessary command: `qterminal -e "bash -c '~/Music/aplay_sound.sh; exec bash'"`, but I don't like using this trick. 

I want to know the reason for the problem and a clear solution.

Any advice in this regard would highly be appreciated.
Best wishes,
Rajib

**Comment by stefonarch** [comment]
Made a  test here, command line `aplay  /home/stef/bin/file.wav` works fine in  ktimer, `aplay  ~/bin/file.wav` does not.

**Reply by tsujan** [reply]
Command lines don't have anything to do with the DE.

**Comment by tsujan** [comment]
If you mean this
![ktimer](https://github.com/user-attachments/assets/b0a2ec00-c1a4-4a3e-96b7-0674bb088210)
It's a weird design flaw in KTimer, because it puts a command-line inside that text entry and pretends that the user could do something with it in the file dialog. *File dialogs are for files, not commands.*

If it works fine under KDE, then it may have some KDE-specific codes, which don't work outside KDE. Again, a problem in the app.

Sorry, I don't have the slightest motivation to read the code of KTimer and find its problems.

**Reply by rajibando** [reply]
Thank you, Mr. stefonarch and Mr. Tsujan, for your own tests and suggestions. Yes, I do agree that the textbox for running a command is coupled with "`Open File` dialog" is weird.

In which case please suggest me a better Timer that has the capability to run bash command lines as well as the terminal. And no weird mixing of the `Command line` with `'File open' dialog`.

As I have stated on my post in KDE forum at the above-mentioned link, I already have a workaround to the current problem by having the Command line open a terminal and then execute the necessary command: `qterminal -e "bash -c '~/Music/aplay_sound.sh; exec bash'"`, but I don’t like using this trick.

**Reply by stefonarch** [reply]
As I said: it looks like ktimer can't "translate" ~ so only full path is working. There is still `sleep [number] && command`  if needed.

**Reply by tsujan** [reply]
> here is still sleep [number] && command if needed.

Yes, IMO, that's the lightest and most flexible method.

For example, I have a shortcut for this script when I need an alarm after X minutes:

```sh
#!/bin/bash

min=`qarma --title="Alarm" --entry --window-icon=time --text="Enter the minutes to wait for alarm:"`
if [ $? -eq 0 ]; then
  case $min in
    *[!0-9]*)
      case $min in
        *[!A-Za-z]*)
          qarma --error --title="Error" --text='You should enter just an integer.' &
          ;;
        *)
          qarma --error --title="Error" --text='You should enter an integer.' &
          ;;
      esac
      ;;
    *)
      if [[ $min -eq 0 ]] || [[ $min -eq "" ]]; then
        qarma --error --title="Error" --text='Zero is not acceptable.' &
        mpg123 -b 1024 -q ~/.Sounds/Finish.mp3 > /dev/null 2>&1 &
      else
        let sec=60*min
        sleep $sec
        mpg123 -b 1024 -q --loop -1 ~/.Sounds/Charpentier.mp3 > /dev/null 2>&1 &
        qarma --info --title="Alarm" --window-icon=time --text='Click OK to stop the alarm.'
        killall mpg123
      fi
      ;;
  esac
else
  mpg123 -b 1024 -q ~/.Sounds/Finish.mp3 > /dev/null 2>&1 &
fi
exit
```

**Reply by rajibando** [reply]
Dear Mr. Tsujan, I am overwhelmed by your sharing your script. But there is always a limitation with scripts. Suppose I am jogging. I have a phone call. I have to pause and then attend the call. The GUI makes it easy to pause the timer. For CLI it is difficult to manage.

I have my own script for the purpose:

```
#!/bin/bash
#### You could change the time
#### Will play iteratively until a key is pressed

sleep 1200  # Waits for 1200 seconds (20 minutes)

# Start loop
while true; do
    aplay ~/Music/YourTimeIsUp.wav  # Play the audio file
    echo "Press any key to stop..." 
    read -t 1 -n 1 key && break  # Wait for any key press to break the loop
done`
```

But the code needs more complexity for the timer to be paused for purpose as stated above, by pressing say the `space bar` key, which could be beyond my cognitive domain. Perhaps I have told you in one of our discussions earlier: that I am not a professional programmer having a programming degree/certificate/diploma. I try it as one of my interests.

BTW, your code has a binary, `qarma`, that isn't in the Debian Repo. Could it be replaced with `zenity`? In fact, I modified the code:

```
#!/bin/bash

min=$(zenity --entry --title="Alarm" --text="Enter the minutes to wait for alarm:" 2>/dev/null)

if [ $? -eq 0 ]; then
  case $min in
    *[!0-9]*)
      case $min in
        *[!A-Za-z]*)
          zenity --error --title="Error" --text='You should enter just an integer.' &
          ;;
        *)
          zenity --error --title="Error" --text='You should enter an integer.' &
          ;;
      esac
      ;;
    *)
      if [[ $min -eq 0 ]] || [[ $min -eq "" ]]; then
        zenity --error --title="Error" --text='Zero is not acceptable.' &
        aplay ~/Music/20minTimeUp.wav > /dev/null 2>&1 &
      else
        let sec=60*min
        sleep $sec
        aplay ~/Music/20minTimeUp.wav > /dev/null 2>&1 &
        zenity --entry --title="Alarm" --text="Enter the minutes to wait for alarm:" 2>/dev/null
        pkill -f "aplay"
      fi
      ;;
  esac
else
  aplay ~/Music/10minTimeUp.wav > /dev/null 2>&1 &
fi
exit
```

Runs. But not as intended. I am presently tweaking it to make it robust.

If considered in depth, a GUI interactive system is far better than a CLI system for some purposes. The `ktimer` is one such.

Thank you also, Mr. stefonarch, for your sharing your code, but I hope I have also addressed the limiation of your code in my above explanation to Mr. Tsujan.

**Reply by tsujan** [reply]
Personally, I'm not interested in GUI alarms, but writing a good Qt-based app that works like a timer/alarm in the systray should be relatively easy, without any KDE dependency whatsoever.

My wild guess is that KTimer may be very old and not maintained actively.

**Reply by rajibando** [reply]
As illustrated in my earlier posts, a GUI countdown timer helps. All the more beneficial if the queerness of `ktimer` is absent.
If a competent programmer takes up the project, beginning from the source code of KDE, all the better. There aren't many timers in Debian.

**Reply by tsujan** [reply]
> If a competent programmer takes up the project, beginning from the source code of KDE, all the better.

No, those KDE dependencies are totally redundant now — perhaps they were needed with Qt3, but now Qt has more than enough tools. A Qt app could be written from scratch.

**Reply by rajibando** [reply]
Great! But will have to wait for a developer with free time and the will to serve the FOSS ecosystem.

I talked about using the Source Code of KDE because one wouldn't have to begin from a blank page. If that isn't a good idea, well then it will be good to drop the source code and begin afresh.

I have observed some queerness regarding development in the FOSS ecosystem. For instance, in the F-Droid repo this discussion could be looked at: [Need for a Reminder/Timer/Scheduler/Organiser which could be repeated as hourly, weekly, monthly, yearly, etc., reminders too, with audio Alarm](https://forum.f-droid.org/t/need-for-a-reminder-timer-scheduler-organiser-which-could-be-repeated-as-hourly-weekly-monthly-yearly-etc-reminders-too-with-audio-alarm/29608)

There are some packages, but no continual alarm accompanying any of them. It appears to be a lack of common sense?! A phone sitting away from me could easily draw my attention by ringing continually. How can a screen alert me? I have to remember to keep looking at the screen from time to time. I may easily lose track.

To me it appears that by dropping off the payment system and adopting the donation system FOSS has cut off a good source of crowdfunding to promote sustained and sustainable development. Money isn't evil. It is just a tool. Could be put to good use. Could be misused. It is the fault of the individuals who misuse, not of money. I wish to pay for software. I don't have much faith in donations. One ensures a professional relationship, strengthens the Social Contract and accountability, the other doesn't.
